### PR TITLE
feat(ui): Add warning signpost for old snaps

### LINKED ISSUE #2988: Consider signposting old snaps?
### Expected behaviour

As a user, I expect snaps in the store to be updated so users get the latest updates, security fixes and features. Some snaps are very old, having been published years ago, and never updated. 

As a developer, I don't have a lot of time to update all my software everywhere, so may forget to update a package in the Snap Store. 

Consider adding some wording to the storefront which tells users that the package is old. 

"This snap in this channel hasn't been updated since February 2017. Contact the developer to ask for an update!"

e.g. https://snapcraft.io/ohmygiraffe

### PR DESCRIPTION
**Pull request by gajeshbhat**
## Done
Adds a warning notification on snap detail pages for snaps that haven't been updated in over 2 years, helping users identify potentially outdated or unmaintained snaps.

Fixes #2988

## Changes

- **Backend Logic**: Added `is_snap_old()` function to detect snaps older than 2 years
- **Date Formatting**: Added `convert_date_month_year()` for cleaner "Month Year" format
- **UI Component**: Added warning notification in snap details sidebar with orange left border
- **Styling**: Custom CSS for clean notification appearance (no borders except orange left border)
- **Testing**: Comprehensive test coverage for date logic and edge cases

## Implementation Details

- **Threshold**: Configurable (default: 2 years)
- **Message**: "This snap in this channel hasn't been updated since [Month Year]. Contact the developer to ask for an update."
- **Positioning**: Appears after "Last Updated" section in sidebar
- **Styling**: Uses Vanilla CSS `p-notification--caution` with custom `.old-snap-warning` class

### Unit Testing
- All existing tests pass
- New tests added for `is_snap_old()` function
- New tests added for `convert_date_month_year()` function
- Manual testing completed on local development server

### Prerequisites
- Checkout to my Branch: `feat/signpost-old-snaps-2988`
- Run the development server locally: `dotrun`

####  Snaps that SHOULD show the warning (2+ years old):
- Visit `http://localhost:8004/mumble` - Should show a warning
- Visit `http://localhost:8004/hello` - Should show a warning  
- Visit `http://localhost:8004/wekan` - Should show a warning

#### Snaps that should NOT show the warning (recently updated):
- Visit `http://localhost:8004/vault` - Should NOT show warning
- Visit `http://localhost:8004/microk8s` - Should NOT show warning

### What to Look For

1. **Warning Appearance**: 
   - Orange left border (no other borders)
   - Warning icon
   - Clean, minimal styling
   - Horizontal line below for consistency

2. **Message Format**:
   - "This snap in this channel hasn't been updated since [Month Year]. Contact the developer to ask for an update."
   - Month and year only (no specific day)

3. **Positioning**:
   - Appears in right sidebar
   - After "Last Updated" section
   - Before "Contact" section (if present)

#### Edge Cases to Test
- Try snaps with different update dates
- Verify warning only appears for snaps 2+ years old
- Check that recent snaps don't show the warning

#### Warning displayed on old snaps (mumble, hello, wekan)

Screenshot from 2025-09-05 23-26-31
Screenshot from 2025-09-05 23-27-05
Screenshot from 2025-09-05 23-27-23

#### No warning on recently updated snaps (vault, microk8s)

Screenshot from 2025-09-05 23-27-51
Screenshot from 2025-09-05 23-28-06

## Notes
- The feature is backward compatible and doesn't affect existing functionality
- Uses timezone-aware datetime functions to avoid deprecation warnings
- Follows existing design patterns and styling conventions
- The message can be easily customized by modifying the template

### CODE REVIEW

**Comment by bartaz** [comment]
Thanks again @gajeshbhat. I'll pass it through copilot first and we'll add it to our maintenance review list. 

With feature requests like this, we usually get them through UX first, before even starting implementation. While your suggestion is OK, it hasn't been decided where such notification should be added, what should it say, how old snap needs to be for it to show up.

That said, we can use your PR to discuss these with UX during review.

File: static/sass/styles.scss [hunk 1]
```diff
@@ -273,6 +273,12 @@ dl {
   }
 }
 
+.old-snap-warning {
+  border: none !important;
```
Comment (bartaz) [inline static/sass/styles.scss]: Not sure what's the reasoning behind this. Ideally we would not want to modify existing Vanilla patterns like that. Especially the use of `!important` is a bad practice in CSS.

Notification does have a "borderless" variant if you were aiming at something like that: https://vanillaframework.io/docs/patterns/notification#appearance

Comment (gajeshbhat) [reply static/sass/styles.scss]: Thank you. I will look into the variant in the docs linked above. I was trying to style the box. My CSS knowledge is definitely below average so I will research and update.

**Comment by gajeshbhat** [comment]
> Thanks again @gajeshbhat. I'll pass it through copilot first and we'll add it to our maintenance review list.
> 
> With feature requests like this, we usually get them through UX first, before even starting implementation. While your suggestion is OK, it hasn't been decided where such notification should be added, what should it say, how old snap needs to be for it to show up.
> 
> That said, we can use your PR to discuss these with UX during review.

Thanks for taking a look @bartaz . Yes, I do remember seeing the `Dev Ready` label on some issues where the UX had been discussed, but as this is a minor change, I assumed that could be done during the PR review. Apologies for not checking early. Happy to work with UX to decide the place, the message and duration to display the warning. Will resolve other comments and wait for the suggestions on this. Thanks for pointing it out. Cheers.

File: webapp/store/logic.py [hunk 2]
```diff
@@ -192,16 +192,76 @@ def convert_date(date_to_convert):
...
+        if date_parsed.tzinfo is None:
+            date_parsed = date_parsed.replace(tzinfo=datetime.timezone.utc)
+
+        now = datetime.datetime.now(datetime.timezone.utc)
+
+        days_diff = (now - date_parsed).days
+        years_diff = days_diff / 365.25  # Account for leap years
+        years_since_update = int(years_diff)
+
+        is_old = days_diff >= (old_threshold_years * 365)
```
[the remark this answers is not in the document]
Comment (gajeshbhat) [reply webapp/store/logic.py]: Updated! Changed to use (old_threshold_years * 365.25) for consistency with the leap year calculation on line 248. Updated tests accordingly to handle the more precise calculation.

File: static/sass/styles.scss (hunk 1 above)
Comment (gajeshbhat) [reply static/sass/styles.scss]: Thank you for the link. Removed the custom CSS with `!important` and switched to using Vanilla Framework's  borderless variant: `p-notification--caution is-borderless`

Comment (gajeshbhat) [reply static/sass/styles.scss]: Kindly note that the orange line on the left in the screenshot is gone, as `is-borderless` class removes all borders. I will update the screenshots.  I think we can leave it like that as the caution symbol still conveys the message and looks cleaner. I will wait for comments from UX before I make any other modifications.

Screenshot from 2025-09-05 23-17-40

**Comment by gajeshbhat** [comment]
Hey @bartaz , Hope you have been well. Any chance we can get a review from UI/UX for this PR ?

**Comment by gajeshbhat** [comment]
Retriggerd CI using `git commit --amend --no-edit && git push --force`and the Inclusive naming check passes now.

**Comment by gajeshbhat** [comment]
@bartaz @edisile Any chance UI/UX can review this ? Happy to answer questions or help update the PR if required.

File: static/sass/styles.scss [hunk 3]
```diff
@@ -277,6 +277,7 @@ dl {
   }
 }
 
+
```
Comment (edisile) [inline static/sass/styles.scss]: is this a mistake? or is there some missing change here?

File: templates/store/snap-details/_details.html [hunk 4]
```diff
@@ -19,6 +19,17 @@ <h3 class="p-muted-heading">Last updated</h3>
   <hr class="p-rule--muted">
 {% endif %}
 
+{% if old_snap_info and old_snap_info.is_old %}
+  <div class="p-notification--caution is-borderless">
+    <div class="p-notification__content">
+      <p class="p-notification__message">
+        This snap in this channel hasn't been updated since {{ old_snap_info.last_updated_formatted }}. Contact the developer to ask for an update.
```
Comment (edisile) [inline templates/store/snap-details/_details.html]: A couple of things with regard to the message:
- the wording doesn't make much sense because the user can't know which channel "this" refers to. sure, they can cross-reference the dates, but what if a snap has multiple channels with the same release date?
- telling users to contact the developer feels a bit pushy, it could be mistaken for encouraging harassment; I'd drop that part

We don't have a proper UX-approved copy, but I think something along the lines of "This snap hasn't been updated in a while, it might be unmaintained and have stability or security issues" would do. As for when to show this message, I have some thoughts I left in another comment

File: webapp/store/snap_details_views.py [hunk 5]
```diff
@@ -191,6 +191,7 @@ def _get_context_snap_details(snap_name, supported_architectures=None):
             "filesize": humanize.naturalsize(binary_filesize),
             "last_updated": logic.convert_date(last_updated),
             "last_updated_raw": last_updated,
+            "old_snap_info": logic.is_snap_old(last_updated),
```
Comment (edisile) [inline webapp/store/snap_details_views.py]: I think computing `is_old` this way is a bit wrong:
1. some developers might push updates to channels that aren't default/stable and leave that one behind on purpose
2. the `get_last_updated_version` function (unlike what the name suggests) doesn't select the most recently updated channel, but rather it looks for a stable channel and selects that one if it's available; this means that snaps that are actively maintained but haven't received updates on the stable channel would be wrongly labeled as "old"

Rather than using `last_updated`, a more correct approach would be using the `updates` tuple:
`updates[0]` is the same channel as `last_updated["channel"]` while `updates[1]` is the most recently updated channel that isn't stable; if the most recent of the two is older than the threshold we set, then we display the message

File: templates/store/snap-details/_details.html [hunk 6]
```diff
@@ -19,6 +19,17 @@ <h3 class="p-muted-heading">Last updated</h3>
   <hr class="p-rule--muted">
 {% endif %}
 
+{% if old_snap_info and old_snap_info.is_old %}
```
Comment (edisile) [inline templates/store/snap-details/_details.html]: This message is related to the release dates that sit in the block above, it would make more sense for this `if` block be near the update dates. Please move it under the dates and above the `hr.p-rule--muted` and remove the `hr` at the end of this block

File: webapp/store/logic.py [hunk 7]
```diff
@@ -193,16 +193,76 @@ def convert_date(date_to_convert):
...
+        return {"is_old": False, "years_since_update": 0}
+
+    try:
+        date_parsed = parser.parse(last_updated_date)
+        if date_parsed.tzinfo is None:
+            date_parsed = date_parsed.replace(tzinfo=datetime.timezone.utc)
+
+        now = datetime.datetime.now(datetime.timezone.utc)
+
+        days_diff = (now - date_parsed).days
```
Comment (edisile) [inline webapp/store/logic.py]: Why do all of this date arithmetic manually? `dateutil.relativedelta(now, date).years >= threshold` is much simpler

File: webapp/store/logic.py [hunk 8]
```diff
@@ -193,16 +193,76 @@ def convert_date(date_to_convert):
...
-    local_timezone = datetime.datetime.utcnow().tzinfo
-    date_parsed = parser.parse(date_to_convert).replace(tzinfo=local_timezone)
-    delta = datetime.datetime.utcnow() - datetime.timedelta(days=1)
+    date_parsed = parser.parse(date_to_convert)
+    # Ensure we have a timezone-aware datetime
+    if date_parsed.tzinfo is None:
+        date_parsed = date_parsed.replace(tzinfo=datetime.timezone.utc)
+
+    now = datetime.datetime.now(datetime.timezone.utc)
+    delta = now - datetime.timedelta(days=1)
```
Comment (edisile) [inline webapp/store/logic.py]: Why is this change necessary? I think this already works correctly

File: static/sass/styles.scss (hunk 3 above)
Comment (gajeshbhat) [reply static/sass/styles.scss]: Yes. Thank you for pointing it out. Fixed.

File: webapp/store/logic.py (hunk 8 above)
Comment (gajeshbhat) [reply webapp/store/logic.py]: It's been a while since I implemented this :sweat_smile: But I believe I was trying to simplify it further. But I have since reverted the change as it works.

File: webapp/store/logic.py (hunk 7 above)
Comment (gajeshbhat) [reply webapp/store/logic.py]: Thanks! `relativedelta` is better than my DIY implementation. Switched.

File: templates/store/snap-details/_details.html (hunk 6 above)
Comment (gajeshbhat) [reply templates/store/snap-details/_details.html]: Moved inside the `{% if updates %}` block as suggested. I see that the layout of the Snap details page has changed (left aligned) since I wrote this and the current position looks better. Thanks.

File: webapp/store/snap_details_views.py (hunk 5 above)
Comment (gajeshbhat) [reply webapp/store/snap_details_views.py]: That's a good point @edisile . I hadn't thought of this case. I updated the implementation to use tuples. Let me know what you think

File: templates/store/snap-details/_details.html (hunk 4 above)
Comment (gajeshbhat) [reply templates/store/snap-details/_details.html]: Thank you for getting that feedback for me from UX. I updated the language to be more informative than pushy!

**Comment by gajeshbhat** [comment]
> Hi there! Thanks for reaching out and sorry for taking so long to reply. Unfortunately we couldn't get a proper formal UX review, but we received some informal comments I reported here along with a few things I noticed

Thank you @edisile for taking the time to review this PR. I have made some changes based on the review feedback.  As changes are not of larger scope, I think a minor UX comment should suffice instead of the full review. Please let me know if you have any question. Cheers.

1. CI check for test-python is failing. I ran this test locally on upstream/main and it is failing as well. Let me know if you folks fix it, I will rebase/
2. Inclusive naming check fails intermittently. I have raised a ticket for this before on this repo (https://github.com/canonical/snapcraft.io/issues/5531#issuecomment-3858788758) and the original repo that hosts this check
3. Black linter was throwing errors (https://github.com/canonical/snapcraft.io/actions/runs/21775757077/job/62831944315) on CI (after rebase, on files I did not touch), I ran the linter locally and updated the lint checks and now its passing.

On a side note, I also have similar smaller UI feature PR that's been waiting for a while now, if you can kindly take a look, that would be helpful : https://github.com/canonical/snapcraft.io/pull/5353 . Thanks again and Cheers.

File: webapp/store/logic.py (hunk 8 above)
Comment (edisile) [reply webapp/store/logic.py]: Not sure how adding more code would simplify it 😂 but fair enough

**Comment by edisile** [comment]
@gajeshbhat thank you! The tests should work now if you rebase on main

File: templates/store/snap-details/_details.html [hunk 9]
```diff
@@ -16,6 +16,15 @@ <h3 class="p-muted-heading">Last updated</h3>
       <li>{{ updates[1]["released-at-display"] }} - <small>{{ updates[1]["track"] }}/{{ updates[1]["risk"] }}</small></li>
     {% endif %}
   </ul>
+  {% if old_snap_info and old_snap_info.is_old %}
+  <div class="p-notification--caution is-borderless">
+    <div class="p-notification__content">
+      <p class="p-notification__message">
+        This snap hasn't been updated since {{ old_snap_info.last_updated_formatted }}. It might be unmaintained and have stability or security issues.
```
Comment (edisile) [inline templates/store/snap-details/_details.html]: As pointed out by UX, we don't need the specific date here. We can keep it vague like "hasn't been updated in a while"

File: webapp/store/logic.py [hunk 10]
```diff
@@ -203,6 +204,61 @@ def convert_date(date_to_convert):
...
+    :returns: Month and year only
+    """
+    date_parsed = parser.parse(date_to_convert)
+    if date_parsed.tzinfo is None:
+        date_parsed = date_parsed.replace(tzinfo=datetime.timezone.utc)
+
+    return date_parsed.strftime("%B %Y")
+
+
+def is_snap_old(last_updated_date, old_threshold_years=2.0):
```
Comment (edisile) [inline webapp/store/logic.py:207]: Since we're removing the update date in the message we can simplify this function, remove the one above and its associated test

File: webapp/store/logic.py [hunk 11]
```diff
@@ -203,6 +204,61 @@ def convert_date(date_to_convert):
...
+        now = datetime.datetime.now(datetime.timezone.utc)
+
+        delta = relativedelta(now, date_parsed)
+        years_since_update = delta.years
+
+        is_old = years_since_update >= old_threshold_years
+
+        return {
+            "is_old": is_old,
+            "years_since_update": years_since_update,
```
Comment (edisile) [inline webapp/store/logic.py]: We're not using `years_since_update` anywhere outside of the tests, I wonder if it would be useful in the long run or we should just remove it

File: tests/store/tests_public_logic.py [hunk 12]
```diff
@@ -482,3 +482,99 @@ def test_get_last_updated_versions(self):
...
+    def test_is_snap_old_recent_snap(self):
+        """Test that a recently updated snap is not considered old"""
+        recent_date = "2023-06-01T10:00:00Z"
+        result = logic.is_snap_old(recent_date)
+        self.assertFalse(result["is_old"])
+        self.assertEqual(result["years_since_update"], 0)
+        self.assertIn("last_updated_formatted", result)
+
+    @freeze_time("2024-01-02")
+    def test_is_snap_old_exactly_two_years(self):
```
Comment (edisile) [inline tests/store/tests_public_logic.py:507]: There's a small inconsistency, test is called `exactly_two_years` but the dates are more than 2 years apart. We should either rename the test or or change the date to match the name

**Review by edisile — CHANGES_REQUESTED** [review]

File: templates/store/snap-details/_details.html (hunk 9 above)
Comment (gajeshbhat) [reply templates/store/snap-details/_details.html]: Done updated.

File: webapp/store/logic.py (hunk 11 above)
Comment (gajeshbhat) [reply webapp/store/logic.py]: I removed it. Updated.

File: tests/store/tests_public_logic.py (hunk 12 above)
Comment (gajeshbhat) [reply tests/store/tests_public_logic.py:507]: Fixed the date to match the name.

File: webapp/store/logic.py (hunk 10 above)
Comment (gajeshbhat) [reply webapp/store/logic.py:207]: Changed to returning simple boolean True/False. Let me know what you think.

**Review by edisile — APPROVED** [review]
Looks good!

**Comment by gajeshbhat** [comment]
> Looks good!

Thank you.

**Review by abbiesims — APPROVED** [review]
